### PR TITLE
Add ExternalAccess API for interceptors

### PR DESCRIPTION
**Pull request by RikkiGibson**
After looking at it for a bit, I thought exposing an EA API would be simpler than trying to paste in some solution in the ILLinker analyzer which visits all the relevant attributes and decides what they mean.

I'm doing something wrong with InternalsVisibleTo right now, I think, the test is refusing to build :( Let's figure that out.

@cston @jcouv @sbomer for review.

Before merge we would of course like to try it out manually and make sure it is fit for purpose on the NativeAOT side.

### CODE REVIEW

File: src/Tools/ExternalAccess/NativeAOT/Internal/InterceptorUtils.cs [hunk 1]
```diff
@@ -0,0 +1,35 @@
...
+    /// <summary>
+    /// If the call denoted by <paramref name="invocation"/> is intercepted in <paramref name="compilation"/>, returns the method symbol for the interceptor.
+    /// Otherwise, returns null.
+    /// </summary>
+    /// <remarks>
+    /// Calling this method forces completion of attributes on all source symbols.
+    /// If using this API in an analyzer, you may want to only access it through
+    /// the callback to 'RegisterCompilationEndAction', which will avoid potential negative perf impact in the IDE.
+    /// </remarks>
+    internal static MethodSymbol? GetInterceptor(InvocationExpressionSyntax invocation, CSharpCompilation compilation, CancellationToken cancellationToken)
```
[the remark this answers is not in the document]
Comment (RikkiGibson) [reply src/Tools/ExternalAccess/NativeAOT/Internal/InterceptorUtils.cs:21]: Yeah, here's the sample I wrote for the api review mail.

```cs
public static void AnalyzeCall(IInvocationOperation invocation)
{
    // native AOT analyzer needs to perform its checks based on the "interceptor" method, not the original method.
    if (invocation is { Syntax: InvocationExpressionSyntax invocationExpression, SemanticModel.Compilation: CSharpCompilation csCompilation })
    {
        // use a possible EA API to solve it?
        var interceptor = InterceptorUtils.GetInterceptor(invocationExpression, csCompilation, cancellationToken: default);
    }
}
```

My thinking was, if the consumer wasn't IOperation-based but still wanted this info, it would still be possible to get the interceptor this way.

[the remark this answers is not in the document]
Comment (captainsafia) [reply src/Tools/ExternalAccess/NativeAOT/Internal/InterceptorUtils.cs:21]: Is resolving the InvocationExpressionSyntax via `IInvocationOperation.Syntax` sufficient here?

File: src/Tools/ExternalAccess/NativeAOT/ExternalAccessNativeAOTResources.resx
Comment (jjonescz) [inline src/Tools/ExternalAccess/NativeAOT/ExternalAccessNativeAOTResources.resx:1]: Is this file needed?

File: src/Tools/ExternalAccess/NativeAOT/Microsoft.CodeAnalysis.ExternalAccess.NativeAOT.csproj [hunk 2]
```diff
@@ -0,0 +1,42 @@
...
+    <PublicAPI Include="PublicAPI.Unshipped.txt" />
+    <PublicAPI Include="InternalAPI.Shipped.txt" />
+    <PublicAPI Include="InternalAPI.Unshipped.txt" />
+  </ItemGroup>
+
+  <ItemGroup>
+    <EmbeddedResource Update="ExternalAccessAspNetCoreResources.resx">
+      <Generator>ResXFileCodeGenerator</Generator>
+      <LastGenOutput>ExternalAccessAspNetCoreResources.Designer.cs</LastGenOutput>
+    </EmbeddedResource>
```
Comment (jjonescz) [inline src/Tools/ExternalAccess/NativeAOT/Microsoft.CodeAnalysis.ExternalAccess.NativeAOT.csproj:39]:
```suggestion
    <EmbeddedResource Update="ExternalAccessNativeAOTResources.resx">
      <Generator>ResXFileCodeGenerator</Generator>
      <LastGenOutput>ExternalAccessNativeAOTResources.Designer.cs</LastGenOutput>
    </EmbeddedResource>
```

File: src/Tools/ExternalAccess/NativeAOT/Internal/InterceptorUtils.cs [hunk 3]
```diff
@@ -0,0 +1,35 @@
+﻿// Licensed to the .NET Foundation under one or more agreements.
+// The .NET Foundation licenses this file to you under the MIT license.
+// See the LICENSE file in the project root for more information.
+
+using System.Threading;
+using Microsoft.CodeAnalysis.CSharp;
+using Microsoft.CodeAnalysis.CSharp.Symbols;
+using Microsoft.CodeAnalysis.CSharp.Syntax;
+
+internal static class InterceptorUtils
```
Comment (333fred) [inline src/Tools/ExternalAccess/NativeAOT/Internal/InterceptorUtils.cs:10]: `Internal` in an EA package is for backing code that the consuming location _shouldn't_ consume.

File: src/Tools/ExternalAccess/NativeAOT/Internal/InterceptorUtils.cs (hunk 1 above)
Comment (333fred) [inline src/Tools/ExternalAccess/NativeAOT/Internal/InterceptorUtils.cs:21]: I think this should be on a `SemanticModel`, not a `Compilation`. It may just call through to a compilation, but conceptually you're asking for semantic information about a syntax node. That's the job of a semantic model.

File: src/Tools/ExternalAccess/NativeAOT/Internal/InterceptorUtils.cs (hunk 3 above)
Comment (jjonescz) [inline src/Tools/ExternalAccess/NativeAOT/Internal/InterceptorUtils.cs:10]: Should this class be in a namespace?

**Comment by RikkiGibson** [comment]
Moving back to draft as we're going to meet offline to discuss how exactly we want to move forward with this.
